File: PumaRhino/ReportAttribute.cpp
/**
 * Puma - CityEngine Plugin for Rhinoceros
 *
 * See https://esri.github.io/cityengine/puma for documentation.
 *
 * Copyright (c) 2021 Esri R&D Center Zurich
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "ReportAttribute.h"

#include "RhinoCallbacks.h"

#include <algorithm>

namespace {

void syncReports(const int shapeID, std::vector<Reporting::ReportAttribute>& reports) {
	size_t currID = reports.size();

	while (currID < shapeID) {
		reports.push_back(Reporting::getEmptyReport(currID));
		currID++;
	}
}

void addToMap(Reporting::GroupedReportMap& reports, const Reporting::ReportAttribute& report,
              const int initialShapeIndex) {
	auto it = reports.find(report.mReportName);
	if (it == reports.end()) {
		std::vector<Reporting::ReportAttribute> newVect;
		syncReports(initialShapeIndex, newVect);
		newVect.push_back(report);

		reports.insert(it, std::make_pair(report.mReportName, newVect));
	}
	else {
		syncReports(initialShapeIndex, reports.at(report.mReportName));
		reports.at(report.mReportName).push_back(report);
	}
}

} // namespace

namespace Reporting {

ReportAttribute getEmptyReport(size_t initialShapeIndex) {
	ReportAttribute report;
	report.mInitialShapeIndex = initialShapeIndex;
	report.mReportName = L"Empty Report";
	report.mType = prt::AttributeMap::PrimitiveType::PT_UNDEFINED;

	return report;
}

void extractReports(size_t initShapeId, Model& model, const prtx::PRTUtils::AttributeMapPtr reports) {
	size_t keyCount = 0;
	auto keys = reports->getKeys(&keyCount);

	for (size_t i = 0; i < keyCount; ++i) {
		auto key = keys[i];

		ReportAttribute ra;
		ra.mInitialShapeIndex = initShapeId;
		ra.mReportName.assign(key);
		ra.mType = reports->getType(key);

		switch (ra.mType) {
			case prt::AttributeMap::PrimitiveType::PT_BOOL:
				ra.mBoolReport = reports->getBool(key);
				break;
			case prt::AttributeMap::PrimitiveType::PT_STRING:
				ra.mStringReport.assign(reports->getString(key));
				break;
			case prt::AttributeMap::PrimitiveType::PT_FLOAT:
				ra.mDoubleReport = reports->getFloat(key);
				break;
			case prt::AttributeMap::PrimitiveType::PT_INT:
				ra.mIntReport = reports->getInt(key);
				break;
			default:
				LOG_ERR << L"Type of report not supported.";
				return;
		}

		// model.mReports.emplace(ra.mReportName, std::move(ra));
		model.addReport(ra);
	}
}

ReportsVector ToReportsVector(const ReportMap& reports) {
	ReportsVector rep;
	rep.reserve(reports.size());

	for (const auto& report : reports) {
		rep.push_back(report.second);
	}

	// auto fct = [](const std::pair<std::wstring, ReportAttribute> p) { return ReportAttribute(p.second); };

	// auto lastit = std::transform(reports.begin(), reports.end(), rep.end(), fct);

	return rep;
}

void GroupedReports::add(const ReportAttribute& report, const int initialShapeIndex) {
	if (report.mType == prt::AttributeMap::PrimitiveType::PT_BOOL) {
		addToMap(mGroupedBoolReports, report, initialShapeIndex);
	}
	else if (report.mType == prt::AttributeMap::PrimitiveType::PT_FLOAT) {
		addToMap(mGroupedDoubleReports, report, initialShapeIndex);
	}
	else if (report.mType == prt::AttributeMap::PrimitiveType::PT_STRING) {
		addToMap(mGroupedStringReports, report, initialShapeIndex);
	}
}

size_t GroupedReports::getReportCount() const {
	return mGroupedBoolReports.size() + mGroupedDoubleReports.size() + mGroupedStringReports.size();
}

bool GroupedReports::getReportKeys(ON_ClassArray<ON_wString>* pKeysArray, ON_SimpleArray<int>* pKeyTypeArray) const {
	auto getReportInfosFunc = [&pKeysArray, &pKeyTypeArray](auto& it) {
		pKeysArray->Append(ON_wString(it.first.c_str()));
		pKeyTypeArray->Append(it.second.front().mType);
	};

	if (!mGroupedDoubleReports.empty()) {
		std::for_each(mGroupedDoubleReports.begin(), mGroupedDoubleReports.end(), getReportInfosFunc);
	}
	if (!mGroupedStringReports.empty()) {
		std::for_each(mGroupedStringReports.begin(), mGroupedStringReports.end(), getReportInfosFunc);
	}
	if (!mGroupedBoolReports.empty()) {
		std::for_each(mGroupedBoolReports.begin(), mGroupedBoolReports.end(), getReportInfosFunc);
	}

	return true;
}

const std::vector<ReportAttribute>& GroupedReports::getDoubleReports(std::wstring key) const {
	try {
		return mGroupedDoubleReports.at(key);
	}
	catch (const std::out_of_range) {
		return EMPTY_REPORTS;
	}
}

const std::vector<ReportAttribute>& GroupedReports::getBoolReports(std::wstring key) const {
	try {
		return mGroupedBoolReports.at(key);
	}
	catch (const std::out_of_range) {
		return EMPTY_REPORTS;
	}
}

const std::vector<ReportAttribute>& GroupedReports::getStringReports(std::wstring key) const {
	try {
		return mGroupedStringReports.at(key);
	}
	catch (const std::out_of_range) {
		return EMPTY_REPORTS;
	}
}

void GroupedReports::clear() {
	mGroupedBoolReports.clear();
	mGroupedStringReports.clear();
	mGroupedDoubleReports.clear();
}

} // namespace Reporting

File: PumaGrasshopper/ReportAttribute.cs
/**
 * Puma - CityEngine Plugin for Rhinoceros
 *
 * See https://esri.github.io/cityengine/puma for documentation.
 *
 * Copyright (c) 2021 Esri R&D Center Zurich
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using Grasshopper.Kernel;
using Grasshopper.Kernel.Parameters;
using Grasshopper.Kernel.Types;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace PumaGrasshopper
{
    public enum ReportTypes
    {
        PT_PT_UNDEFINED,		///< Undefined type.
		PT_STRING,				///< String type.
		PT_FLOAT,				///< Float type.
		PT_BOOL,				///< Boolean type.
		PT_INT,					///< Integer type.
		PT_STRING_ARRAY,		///< String array type.
		PT_FLOAT_ARRAY,			///< Float array type.
		PT_BOOL_ARRAY,			///< Boolean array type.
		PT_INT_ARRAY,			///< Integer array type.
		PT_BLIND_DATA,			///< Blind data type.
		PT_BLIND_DATA_ARRAY,	///< Blind data array type.
		PT_COUNT
    };

    public class ReportAttribute: GH_Goo<Param_GenericObject>
    {
        private readonly string mKey;
        private readonly ReportTypes mType;

        public override bool IsValid => true;

        public override string TypeName => "ReportAttribute";

        public override string TypeDescription => "Contains a cga report {key : value} where key is the report name and value is either a double, a boolean or a string.";

        public int InitialShapeIndex { get; }
        public string StringValue { get; set; } = "";
        public double DoubleValue { get; set; } = 0;
        public bool BoolValue { get; set; } = false;

        public ReportAttribute(int initialShapeIndex, string key, ReportTypes type)
           : base()
        {
            this.InitialShapeIndex = initialShapeIndex;
            this.mKey = key;
            this.mType = type;
        }

        public string GetFormatedValue()
        {
            switch(mType)
            {
                case ReportTypes.PT_FLOAT:
                    return DoubleValue.ToString();
                case ReportTypes.PT_STRING:
                    return StringValue;
                case ReportTypes.PT_BOOL:
                    return BoolValue.ToString();
                default:
                    return "UNDEFINED";
            }
        }

        public string GetReportKey() { return mKey; }

        public ReportTypes GetReportType() { return mType; }

        public static ReportAttribute CreateReportAttribute(int initialShapeIndex, string name, ReportTypes type, string value)
        {
            ReportAttribute report = new ReportAttribute(initialShapeIndex, name, type)
            {
                StringValue = value
            };
            return report;
        }

        public static ReportAttribute CreateReportAttribute(int initialShapeIndex, string name, ReportTypes type, double value)
        {
            ReportAttribute report = new ReportAttribute(initialShapeIndex, name, type)
            {
                DoubleValue = value
            };
            return report;
        }

        public static ReportAttribute CreateReportAttribute(int initialShapeIndex, string name, ReportTypes type, bool value)
        {
            ReportAttribute report = new ReportAttribute(initialShapeIndex, name, type)
            {
                BoolValue = value
            };
            return report;
        }

        public override IGH_Goo Duplicate()
        {
            return this;
        }

        public override string ToString()
        {
            return "[ InitialShapeIndex: " + InitialShapeIndex + ", Key: " + mKey + ", Value: " + GetFormatedValue() + " ]";
        }

        public string ToNiceString()
        {
            return "  " + mKey + " : " + GetFormatedValue();
        }
    }
}
